src/platform/plugins/shared/embeddable/common/embeddable_content_management/registry.ts

```diff
@@ -7,14 +7,22 @@
  * License v3.0 only", or the "Server Side Public License, v 1".
  */
 
+// eslint-disable-next-line max-classes-per-file
 import { EmbeddableContentManagementDefinition } from '..';
 
-export class EmbeddableContentManagementRegistry {
-  private registry: Map<string, () => Promise<EmbeddableContentManagementDefinition>> = new Map();
+class EmbeddableContentManagementRegistry<IsAsync extends true | false = false> {
+  private registry: Map<
+    string,
+    () => IsAsync extends true
+      ? Promise<EmbeddableContentManagementDefinition>
+      : EmbeddableContentManagementDefinition
+  > = new Map();
 
   public registerContentManagementDefinition = (
     id: string,
-    getDefinition: () => Promise<EmbeddableContentManagementDefinition>
+    getDefinition: () => IsAsync extends true
+      ? Promise<EmbeddableContentManagementDefinition>
+      : EmbeddableContentManagementDefinition
```

Comment:
nit, this function return type is used in a couple of places. should we define the type separately?